File: backend/app/domain/entities/comprobante.py
from dataclasses import dataclass, field
from uuid import UUID, uuid4
from datetime import datetime
from app.domain.enums.estado_comprobante import EstadoComprobante
from app.domain.enums.tipo_comprobante import TipoComprobante
from app.shared.exceptions import DomainError


@dataclass
class Comprobante:
    tenant_id: UUID
    tipo: TipoComprobante
    establecimiento: str
    punto_emision: str
    datos: dict
    id: UUID = field(default_factory=uuid4)
    clave_acceso: str | None = None
    secuencial: str | None = None
    estado: EstadoComprobante = EstadoComprobante.DRAFT
    idempotency_key: str | None = None
    external_reference: str | None = None
    numero_autorizacion: str | None = None
    fecha_autorizacion: datetime | None = None
    s3_key_xml: str | None = None
    s3_key_xml_firmado: str | None = None
    s3_key_xml_autorizado: str | None = None
    s3_key_pdf: str | None = None
    lote_id: UUID | None = None
    retry_count: int = 0
    error_detalle: str | None = None
    created_at: datetime = field(default_factory=datetime.utcnow)
    updated_at: datetime = field(default_factory=datetime.utcnow)

    def transition_to(self, nuevo_estado: EstadoComprobante) -> None:
        valid = EstadoComprobante.VALID_TRANSITIONS.get(self.estado, set())
        if nuevo_estado not in valid:
            raise DomainError(
                f"Transición inválida: {self.estado} → {nuevo_estado}",
                code="INVALID_STATE_TRANSITION",
            )
        self.estado = nuevo_estado
        self.updated_at = datetime.utcnow()

    def increment_retry(self) -> None:
        self.retry_count += 1
        self.updated_at = datetime.utcnow()

    def mark_failed(self, reason: str) -> None:
        self.error_detalle = reason
        self.estado = EstadoComprobante.FAILED
        self.updated_at = datetime.utcnow()

File: frontend/shared/src/types/comprobante.ts
export type TipoComprobante = "01" | "03" | "04" | "05" | "06" | "07";

export type EstadoComprobante =
  | "DRAFT"
  | "PENDING_VALIDATION"
  | "VALIDATION_FAILED"
  | "QUEUED"
  | "PROCESSING"
  | "XML_GENERATED"
  | "XML_SIGNED"
  | "SENT_TO_SRI"
  | "RECEIVED_BY_SRI"
  | "RETURNED_BY_SRI"
  | "PENDING_AUTHORIZATION"
  | "AUTHORIZED"
  | "NOT_AUTHORIZED"
  | "PENDING_CANCELLATION"
  | "CANCELLED"
  | "EMAIL_PENDING"
  | "EMAIL_SENT"
  | "EMAIL_FAILED"
  | "FAILED"
  | "RETRY_PENDING"
  | "MANUAL_REVIEW_REQUIRED";

export interface Comprobante {
  id: string;
  tenant_id: string;
  tipo: TipoComprobante;
  clave_acceso: string | null;
  establecimiento: string;
  punto_emision: string;
  secuencial: string | null;
  estado: EstadoComprobante;
  idempotency_key: string | null;
  external_reference: string | null;
  numero_autorizacion: string | null;
  fecha_autorizacion: string | null;
  retry_count: number;
  error_detalle: string | null;
  created_at: string;
  updated_at: string;
}

export interface CreateComprobanteRequest {
  tipo: TipoComprobante;
  establecimiento: string;
  punto_emision: string;
  idempotency_key?: string;
  external_reference?: string;
  datos: Record<string, unknown>;
}

export interface ComprobanteListResponse {
  items: Comprobante[];
  total: number;
  page: number;
  page_size: number;
}
